### ISSUE TITLE
Suggestion: change TokenKind to be more structured

### ISSUE THREAD

**Issue by MovingtoMars** [yberreby/rgo#2]
Something like this:

```
pub enum TokenKind {
    Ident,
    Delimiter(Delimiter),
    Keyword(Keyword),
    BinaryOperator(BinaryOperator),
    ...
}

pub enum Delimeter {
    LParen,
    ...
}

pub enum Keyword {
    Break,
    ...
}
```

One example of where this would be helpful is the precendence() method. Instead of panicing if the token is not a binop, simple move the method onto the BinaryOperator enum.

**Comment by yberreby** [comment]
I actually [started with a similar approach](https://github.com/yberreby/rgo/blob/2ad246bac56045171eabde7d192c2312800f4c09/src/token.rs#L3-L7), except that I was also storing the value of tokens that had one (e.g. `Ident`).

I initially moved away from this approach because non-lexical borrows forced me to clone token values, but this problem doesn't apply to token _kinds_, since they have no value.

Around the same time (or maybe later, I'm not sure), I discovered [this](https://www.reddit.com/r/rust/comments/4g3yz9/collapsing_nested_enum_tags/). Basically, nested enums take up more space than necessary,

_However_, I just ran [a quick experiment](http://is.gd/MwVdrc), and the problem is definitely not as bad in this particular case. We're going from one byte to two. Because of struct padding, this should result in a 0% increase in memory usage in most cases, so this change might be worth it. Additionally, since nested C-like enums are pretty much the simplest case of nested enums, they should be pretty simple to optimize, so this might be optimized away entirely.

TL;DR: good idea, let's try.

**Comment by MovingtoMars** [comment]
Cool! I'd be interested in implementing this.

**Comment by yberreby** [comment]
Really? Well, that would be great. As I said in #1, I don't have much time to work on rgo till the holidays, so any help is greatly appreciated. Especially since you're already contributing to a [compiler](https://github.com/ark-lang/ark).

**Comment by MovingtoMars** [comment]
I'm a bit burnt out on Ark at the moment (bug ark-lang/ark#709 is hard to solve) and I'm learning Rust so I thought it would be fun to contribute.

**Comment by yberreby** [comment]
> I thought it would be fun to contribute.

In my experience, it is! I've never had so much fun developing software, actually. Rust is very, very good for compiler writing. I've found ADTs (i.e. enums), pattern matching, memory safety, a strong ecosystem and strong typing to be incredibly useful for this project, even though it's very young.

**Comment by yberreby** [comment]
@MovingtoMars I opened [a handful of issues](https://github.com/yberreby/rgo/issues) that you might find interesting if you want to contribute.

**Comment by yberreby** [comment]
I just realized there's another thing to take into account: verbosity.

Compare [this](https://github.com/yberreby/rgo/blob/ad5f3df2a89d874758806e5e45a1cf931257b185/src/lexer/mod.rs#L606-L619):

``` rust
match *t {
    Token::Ident(_) |
    Token::Literal(_) |
    Token::Keyword(Keyword::Break) |
    Token::Keyword(Keyword::Continue) |
    Token::Keyword(Keyword::Fallthrough) |
    Token::Keyword(Keyword::Return) |
    Token::Increment |
    Token::Decrement |
    Token::CloseDelim(DelimToken::Paren) |
    Token::CloseDelim(DelimToken::Bracket) |
    Token::CloseDelim(DelimToken::Brace) => true,
    _ => false,
}
```

To [this](https://github.com/yberreby/rgo/blob/3fda209eb738f8a625dba1795c7a92f3c651e7b4/src/lexer/mod.rs#L606-L619):

``` rust
match t {
    TokenKind::Ident |
    TokenKind::Break |
    TokenKind::Continue |
    TokenKind::Fallthrough |
    TokenKind::Return |
    TokenKind::Increment |
    TokenKind::Decrement |
    TokenKind::RParen |
    TokenKind::RBracket |
    TokenKind::RBrace => true,
    t if t.is_literal() => true,
    _ => false,
}
```

It's not much, but certainly annoying when repeated across the whole codebase.

**Comment by MovingtoMars** [comment]
I've done it like this:

``` rust
    match t {
        TokenKind::Ident |
        TokenKind::Break |
        TokenKind::Continue |
        TokenKind::Fallthrough |
        TokenKind::Return |
        TokenKind::Increment |
        TokenKind::Decrement => true,
        TokenKind::Delim(delim) => match delim {
            Delim::RParen |
            Delim::RBracket |
            Delim::RBrace => true,
            _ => false,
        },
        t if t.is_literal() => true,
        _ => false,
    }
```

**Comment by yberreby** [comment]
That's a solution for matching multiple tokens at once, but doesn't solve single-token matching. Look inside [this old parser/mod.rs](https://github.com/yberreby/rgo/blob/ad5f3df2a89d874758806e5e45a1cf931257b185/src/parser/mod.rs) and compare it to the current one.

Example:

``` rust
self.eat(&Token::OpenDelim(DelimToken::Paren));
```

Versus

``` rust
self.eat(TokenKind::LParen)
```

Again, it's not _much_, but it makes writing tests and working on the parser more tedious.

**Comment by MovingtoMars** [comment]
Now that I'm on to operators, I've come across a problem. My original plan was to have `BinaryOp` and `UnaryOp`, but `+`, `-`, `&` and `*` fit in both categories.

**Comment by yberreby** [comment]
A possible solution would be to convert a `TokenKind` to `{BinaryOp, UnaryOp, ...}` when inserting them into the AST, instead of trying to do that upfront. At that point, we already know whether they are unary or binary.

**Comment by MovingtoMars** [comment]
Is there a reason that all the TokenKind methods take `&self` instead of `self`?

**Comment by yberreby** [comment]
I usually try to avoid depending on a type being `Copy` whenever I can, to ease refactoring. Although in that particular case, it shouldn't matter much.

**Comment by yberreby** [comment]
Fixed by #7. We'll see how this turns out.

**Comment by MovingtoMars** [comment]
So you don't mind it being changed to `self`? I think it's much cleaner this way,

**Comment by yberreby** [comment]
No, I don't. If you want to, then go ahead :)

Since you're contributing, may I ask what your plans are, going forward? I personally won't be able to _seriously_ work on `rgo` until July 2, when my exams end.

**Comment by MovingtoMars** [comment]
Well I think this is an interesting project to work on. (I love compilers). So I'd like to work on a few things for a while. Not sure what after that.

**Comment by yberreby** [comment]
Okay, fine! I gave you push access, so you can merge small changes / fixes directly.

Just _please_ don't use the GitHub "merge" button. It creates an ugly history; rebasing + fast-forward merges are much cleaner. See [this StackOverflow question](http://stackoverflow.com/questions/9994093/automatic-merge-of-pull-requests-on-github-without-the-merge-bubble).

From the first answer:

```
$ git checkout master
$ git remote add cameronmcefee git://github.com/cameronmcefee/Spoon-Knife.git
$ git fetch cameronmcefee 
$ git merge cameronmcefee/my-branch
$ git push origin master
```

GitHub is able to detect merges done this way, and marks the relevant PR as merged automatically.

**Comment by MovingtoMars** [comment]
Ok, got it!

**Comment by yberreby** [comment]
There's another issue with this approach: some tokens can be operators, or have other functions. I'm thinking about the star `*` and the left arrow `<-`, which can be operators _or_ be part of a type.

What about using a flat `TokenKind` when lexing, but storing structured tokens in the AST as the result of parsing, for example, binary expressions?

In addition to solving the previous ambiguity, it would let us reduce verbosity while retaining type safety.

**Comment by MovingtoMars** [comment]
Hmm, do you see that as a big problem? I don't think it's that bad to have `*` be an operator and part of a type. You'd just need to check for `TokenKind::Operator(Operator::Star)` when parsing the type, which seems like a more minor thing.

**Comment by yberreby** [comment]
Not a _big_ problem, but it's a bit inconsistent. When lexing, we don't have enough context to distinguish the various constructs; I think this belongs to the parsing phase, where we _will_ have to match on the `TokenKind` to extract the `Operator` anyway. At this point, there is little benefit in keeping the structured tokens _before_ parsing, in my opinion. Using them when storing tokens in the AST is a good thing, though.

Basically, we'd call `TokenKind::to_operator(&self)`, `to_delim`, etc, when inserting tokens into the AST, which could either return an `Option` or panic if the kind doesn't match.

**Comment by MovingtoMars** [comment]
Can you expand on what you mean by the AST thing?

**Comment by yberreby** [comment]
Sure.

Here's a snippet from the parser:

``` rust
fn parse_potential_binary_expr(&mut self, prec1: i32) -> PResult<ast::Expr> {
    // Grammar:
    //  Expression binary_op Expression
    trace!("parse_potential_binary_expr");

    let mut x = ast::Expr::Unary(try!(self.parse_unary_expr()));

    loop {
        let op_kind = self.token.kind;
        let precedence = op_kind.precedence();
        if precedence < prec1 {
            return Ok(x);
        }

        self.bump();

        let y = try!(self.parse_potential_binary_expr(precedence + 1));
        x = ast::Expr::Binary(ast::BinaryExpr {
            lhs: Box::new(x),
            operator: op_kind,
            rhs: Box::new(y),
        });
    }
}
```

When parsing, we're constructing the AST. `ast::BinaryExpr` currently stores a `TokenKind`. This means that it we lose type safety: it would be possible for us to mistakenly insert, say, a open parenthesis in the `operator` field. Consumers of the AST will have to account for this, and add a default branch to their matches, since there is not guarantee that the `operator` is actually an operator and not something else.

Now, to avoid this, we can store a `token::Operator`. To do this, with the current "structured tokens", we'd do something like:

``` rust
if let TokenKind::Operator(op) = kind {
    // ...insert the operator in BinaryExpr ...
} else {
    panic!("BUG: mismatched token kind; expected an operator")
}
```

We have to do a check _anyway_ to make sure our token is in fact an operator. Since this check is mandatory, my suggestion was that we'd move it into methods belonging to `TokenKind`, to convert a token into a `Delim`, an `Operator`, etc.

**Comment by MovingtoMars** [comment]
I think that, rather than use token kinds to store the operation for a binary expression, there could be `ast::BinaryOperation`, with methods to convert from tokens.

```
pub enum BinaryOperation {
    Add,
    Sub,
    Mul,
    ...
```

**Comment by yberreby** [comment]
Indeed, it would be more appropriate to put that enum in `ast`.

But, as I said, this removes much of the usefulness of "structured token", so at that point, we may as well get rid of them and of the extra verbosity they imply.

**Comment by MovingtoMars** [comment]
I still think it's cleaner to use a nested TokenKind for ease of parsing. But, you make the decision on it.

**Comment by yberreby** [comment]
Can you point me to a place where a nested TokenKind makes parsing substantially easier?

**Comment by MovingtoMars** [comment]
Now that I look into it, having it nested doesn't make code any more concise or anything. It's more of a stylistic choice. I guess I think it could be a good idea to make it flat and go with the enums in `ast`.

**Comment by yberreby** [comment]
~~Nested~~ flat it is, then :)

**Comment by MovingtoMars** [comment]
Wait, so you don't want to change it back?

**Comment by yberreby** [comment]
My bad ahah. I meant _flat_. Sorry :x
